alf/layers_test.py

```diff
@@ -269,6 +271,126 @@ def test_harr_basis_correctness(self, input_size, output_size):
             torch.eye(dec.weight.shape[0]),
             epsilon=1e-6)
 
+    def test_transformer_block_shift(self):
+        n = 32
+        m = 7
+        l = 2 * n - 1
+        b = 10
+        h = 3
+        x = torch.rand(b, h, m, l)
+        y = alf.layers.TransformerBlock._shift(x)
+        for i in range(m):
+            for j in range(n):
+                self.assertEqual(y[:, :, i, j], x[:, :, i, n - 1 + i - j])
+
+    @parameterized.parameters(0, 1, 2)
+    def test_transformer_block(self, task_type):
+        batch_size = 100
+        max_len = 32
+        d_model = 64
+        layers = [alf.layers.FC(4, d_model, torch.relu_)]
+        for i in range(2):
+            layers.append(
+                alf.layers.TransformerBlock(
+                    d_model=d_model,
+                    d_k=d_model,
+                    d_v=d_model,
+                    num_heads=3,
+                    memory_size=max_len,
+                    relative_positional_encoding=task_type == 2))
+        layers.append(alf.layers.FC(d_model, 1))
+        model = nn.Sequential(*layers)
+
+        def _get_batch_content_based(batch_size):
+            """
+            A simple memory task:
+            In the first half of the sequence x[n] (i, j < max_len // 2):
+                one of x[n, i, 0] is set to 1, which means type 0 value is x[n, i, 3]
+                one of x[n, j, 1] is set to 1, which means type 1 value is x[n, j, 3]
```

Comment:
There is no description for dimension [:, :, 2]? To be clearer, "one of x[n, j, 1] (over j) is set to 1..."